planner/core/plan_cache.go

```diff
@@ -90,90 +113,105 @@ func GetPlanFromSessionPlanCache(ctx context.Context, sctx sessionctx.Context, i
+				if !unionScan && tableHasDirtyContent(sctx, tblInfo) {
+					planValid = false
+					// TODO we can inject UnionScan into cached plan to avoid invalidating it, though
+					// rebuilding the filters in UnionScan is pretty trivial.
 					sctx.PreparedPlanCache().Delete(cacheKey)
 					break
 				}
-				if !cachedVal.varTypesUnchanged(binVarTypes, txtVarTypes) {
-					continue
+			}
+			if planValid {
+				err := RebuildPlan4CachedPlan(cachedVal.Plan)
+				if err != nil {
+					logutil.BgLogger().Debug("rebuild range failed", zap.Error(err))
+					return nil, nil, false, nil
 				}
-				planValid := true
-				for tblInfo, unionScan := range cachedVal.TblInfo2UnionScan {
-					if !unionScan && tableHasDirtyContent(sctx, tblInfo) {
-						planValid = false
-						// TODO we can inject UnionScan into cached plan to avoid invalidating it, though
-						// rebuilding the filters in UnionScan is pretty trivial.
-						sctx.PreparedPlanCache().Delete(cacheKey)
-						break
-					}
+				sessVars.FoundInPlanCache = true
+				if len(bindSQL) > 0 {
+					// When the `len(bindSQL) > 0`, it means we use the binding.
+					// So we need to record this.
+					sessVars.FoundInBinding = true
 				}
-				if planValid {
-					err := RebuildPlan4CachedPlan(cachedVal.Plan)
-					if err != nil {
-						logutil.BgLogger().Debug("rebuild range failed", zap.Error(err))
-						goto REBUILD
-					}
-					sessVars.FoundInPlanCache = true
-					if len(bindSQL) > 0 {
-						// When the `len(bindSQL) > 0`, it means we use the binding.
-						// So we need to record this.
-						sessVars.FoundInBinding = true
-					}
-					if metrics.ResettablePlanCacheCounterFortTest {
-						metrics.PlanCacheCounter.WithLabelValues("prepare").Inc()
-					} else {
-						planCacheCounter.Inc()
-					}
-					stmtCtx.SetPlanDigest(preparedStmt.NormalizedPlan, preparedStmt.PlanDigest)
-					return cachedVal.Plan, cachedVal.OutPutNames, nil
+				if metrics.ResettablePlanCacheCounterFortTest {
+					metrics.PlanCacheCounter.WithLabelValues("prepare").Inc()
+				} else {
+					planCacheCounter.Inc()
 				}
-				break
+				stmtCtx.SetPlanDigest(preparedStmt.NormalizedPlan, preparedStmt.PlanDigest)
+				return cachedVal.Plan, cachedVal.OutPutNames, true, nil
 			}
+			break
 		}
 	}
+	return nil, nil, false, nil
+}
+
+func rebuildPhysicalPlan(ctx context.Context, sctx sessionctx.Context, is infoschema.InfoSchema,
+	preparedStmt *CachedPrepareStmt, ignorePlanCache bool, cacheKey kvcache.Key, latestSchemaVersion int64,
+	isBinProtocol bool, varsNum int, binVarTypes []byte, txtVarTypes []*types.FieldType) (plan Plan, names []*types.FieldName, err error) {
+	prepared := preparedStmt.PreparedAst
+	sessVars := sctx.GetSessionVars()
+	stmtCtx := sessVars.StmtCtx
 
-REBUILD:
 	planCacheMissCounter.Inc()
 	stmt := prepared.Stmt
 	p, names, err := OptimizeAstNode(ctx, sctx, stmt, is)
 	if err != nil {
-		return nil, nil, err
+		return p, names, err
```

Comment:
Why change here? The following are same.

Reply:
For Looks unified，i will change it back

Reply:
> Why change here? The following is same.

For Looks unified，i will change it back